### PR TITLE
[Misc] add option to get_distributed_init_method with "env://" 

### PR DESCRIPTION
**Pull request by HoustonJ2013**
This option makes it easier to configure multi-node setup on a cluster with slurm. The following script works on my two nodes slurm system.  

```
#!/bin/sh

HOST_NAME=$(hostname)

export MASTER_ADDR=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | tail -n 1)

echo "Working on ${HOST_NAME}"

start_port=$(echo $SLURM_STEP_RESV_PORTS | sed 's/-/ /g' | awk {'print $1'})

end_port=$(echo $SLURM_STEP_RESV_PORTS | sed 's/-/ /g' | awk {'print $2'})

all_ports=$(seq $start_port $end_port)

ports_array=($all_ports)

server_port=${ports_array[0]}

torch_port=${ports_array[1]}

 

# Ray slurm examples https://docs.ray.io/en/latest/cluster/vms/user-guides/community/slurm.html

ray_port=6379

echo "all ports used "$server_port,$ray_port,$torch_port

 

if [ -z "$GPU_PER_NODE" ]

then

    GPU_PER_NODE=1

fi

 

if [ -z "$N_NODES" ]

then

    N_NODES=2

fi

WORLDSIZE=$(expr $GPU_PER_NODE "*" $N_NODES)

 

# set env variable for torch distributed

export MASTER_PORT=$torch_port

export WORLD_SIZE=$WORLDSIZE

export RANK=$SLURM_PROCID

echo "TORCH env "$MASTER_ADDR,$MASTER_PORT,$WORLD_SIZE,$RANK

 

# set env variable for Ray

export master_ip=$(host $MASTER_ADDR | awk '{print $4}')

 

echo "World size is "$WORLDSIZE

echo "slurm id ${SLURM_PROCID}"

 

if [[ "$MASTER_ADDR" == "$HOST_NAME" ]]

then

    HOST_IP=$HOST_NAME

    echo "Setting HOST_IP on ${HOST_NAME}"

    hostip=$(hostname -I | cut -f1 -d' ')

    echo "Serving LLM at the following address"

    echo $hostip:$server_port   

    echo "Starting a Ray instance .."

    ray start --head --node-ip-address=$master_ip \

    --port=$ray_port --dashboard-host 0.0.0.0 --dashboard-port 8266 \

    --num-cpus $SLURM_CPUS_PER_TASK --num-gpus $GPU_PER_NODE --block &

    sleep 40

    python -m vllm.entrypoints.openai.api_server --model MODEL_PATH  --tensor-parallel-size $WORLDSIZE --dtype float16 --port $server_port

else

    sleep 10

    ray start --address "${master_ip}:${ray_port}" \

        --num-cpus $SLURM_CPUS_PER_TASK --num-gpus $GPU_PER_NODE --block &

    sleep 1d # hold the resource on worker node for ray, exiting the current flow will release the resource

fi
```

### CODE REVIEW

File: vllm/utils.py [hunk 1]
```diff
@@ -212,9 +212,13 @@ def get_ip() -> str:
 
 
 def get_distributed_init_method(ip: str, port: int) -> str:
-    # Brackets are not permitted in ipv4 addresses,
-    # see https://github.com/python/cpython/issues/103848
-    return f"tcp://[{ip}]:{port}" if ":" in ip else f"tcp://{ip}:{port}"
+    if (os.environ.get("MASTER_ADDR") is not None) and \
+        (os.environ.get("MASTER_PORT") is not None) and \
+            (os.environ.get("WORLD_SIZE") is not None) and \
+                (os.environ.get("RANK") is not None):
```
Comment (zhuohan123) [inline vllm/utils.py:218]: Please use `()` for line continuation.
```suggestion
    if ((os.environ.get("MASTER_ADDR") is not None) and
        (os.environ.get("MASTER_PORT") is not None) and
            (os.environ.get("WORLD_SIZE") is not None) and
                (os.environ.get("RANK") is not None)):
```

**Review by zhuohan123 — COMMENTED** [review]
Can you give a more detailed example in which case this will be used? Right now the distributed initialization is handled by ray or python multiprocess. When do we need this?

**Comment by youkaichao** [comment]
@HoustonJ2013 in your case, you only launch python in one node, and inside the script, it will use `ray` to launch tasks in the rest nodes. How does slurm come into play here?

**Comment by youkaichao** [comment]
please see https://github.com/vllm-project/vllm/pull/4014 , you should be able to use `env://` now.

**Pull request closed without merging**